newsrevenuehub/rev-engine
# DEV-2016 | Refactor PaymentProvider to prevent client side errors

Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

This PR is initially into DEV-2031 because it presupposes those changes. Once that PR has been merged, this PR's base should be changed to main.


#### What's this PR do?

Fixes a bug that was causing live donation pages to 500 if their revenue program's payment provider is deleted. Additionally, makes some changes to ensure can't get in that state.

- In Django admin
  - Prevents payment providers from being deleted if there are any donation pages with non-null published_date that depend on the provider (via intermediate RP).  When visiting the admin change view for a such a provider
  - The delete button is hidden, 
  - There is a message at the top of the interface that identifies how many donation pages (with a past or future publication date) depend on it. 
  - That message contains a link to the revenue program admin with a search pre-applied for the payment provider's stripe account id, which filters list by RP's that reference the provider.
- Adds a test demonstrating admin behavior around payment provider deletion
- Adds a test for PaymentProvider model demonstrating the `get_dependent_pages_with_publication_date` business logic.
- Fixes a bug in `apps.pages.helpers.PageFullDetailHelper._validate_live_page_request` that caused an uncaught exception when no payment provider on request live page (which was the acute cause of 500 error that led to this ticket).
- Adds for page detail serializer proving it doesn't break on missing payment provider
- Adds tests to apps.pages.tests.test_viewsets that proves API views don't break when no provider for list and detail view. Adds separate test for live-detail action on same viewset showing that 404 error and message when no provider.


#### Why are we doing this? How does it help us?

- Fixes an acute bug
- Prevents payment providers from being deleted that have live pages because that's a problem.

#### How should this be manually tested?

Set up a published donation page.

In the Django admin, go to the change view for that page's revenue program's payment provider.

There shouldn't be a delete button, and there should be a message at the top of the interface that explains why it can't be deleted, including a link to see RPs filtered by that provider's stripe account id.

<img width="758" alt="Screen Shot 2022-07-11 at 10 50 25 PM" src="https://user-images.githubusercontent.com/960351/178397980-b543f651-dbe9-44c4-b6db-355889f8f16d.png">

#### Have automated unit tests been added?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

N/A

#### What are the relevant tickets? Add a link to any relevant ones.

[DEV-2016](https://news-revenue-hub.atlassian.net/jira/software/c/projects/DEV/boards/3?modal=detail&selectedIssue=DEV-2016)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No

#### Are there next steps? If so, what? Have tickets been opened for them?

N/A
